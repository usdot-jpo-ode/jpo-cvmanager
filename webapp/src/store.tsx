import { Action, ThunkAction, configureStore } from '@reduxjs/toolkit'
import rsuReducer from './generalSlices/rsuSlice'
import userReducer from './generalSlices/userSlice'
import wzdxReducer from './generalSlices/wzdxSlice'
import configReducer from './generalSlices/configSlice'
import intersectionReducer from './generalSlices/intersectionSlice'
import adminAddOrganizationReducer from './features/adminAddOrganization/adminAddOrganizationSlice'
import adminAddRsuReducer from './features/adminAddRsu/adminAddRsuSlice'
import adminAddIntersectionReducer from './features/adminAddIntersection/adminAddIntersectionSlice'
import adminAddUserReducer from './features/adminAddUser/adminAddUserSlice'
import adminEditOrganizationReducer from './features/adminEditOrganization/adminEditOrganizationSlice'
import adminEditRsuReducer from './features/adminEditRsu/adminEditRsuSlice'
import adminEditIntersectionReducer from './features/adminEditIntersection/adminEditIntersectionSlice'
import adminEditUserReducer from './features/adminEditUser/adminEditUserSlice'
import adminOrganizationTabReducer from './features/adminOrganizationTab/adminOrganizationTabSlice'
import adminOrganizationTabUserReducer from './features/adminOrganizationTabUser/adminOrganizationTabUserSlice'
import adminOrganizationTabRsuReducer from './features/adminOrganizationTabRsu/adminOrganizationTabRsuSlice'
import adminOrganizationTabIntersectionReducer from './features/adminOrganizationTabIntersection/adminOrganizationTabIntersectionSlice'
import adminRsuTabReducer from './features/adminRsuTab/adminRsuTabSlice'
import adminIntersectionTabReducer from './features/adminIntersectionTab/adminIntersectionTabSlice'
import adminUserTabReducer from './features/adminUserTab/adminUserTabSlice'
import adminNotificationTabReducer from './features/adminNotificationTab/adminNotificationTabSlice'
import adminAddNotificationReducer from './features/adminAddNotification/adminAddNotificationSlice'
import adminEditNotificationReducer from './features/adminEditNotification/adminEditNotificationSlice'
import menuReducer from './features/menu/menuSlice'
import asn1DecoderSlice from './features/intersections/decoder/asn1-decoder-slice'
import intersectionMapReducer from './features/intersections/map/map-slice'
import intersectionMapLayerStyleReducer from './features/intersections/map/map-layer-style-slice'
import dataSelectorReducer from './features/intersections/data-selector/dataSelectorSlice'
import mapSliceReducer from './pages/mapSlice'

export const setupStore = (preloadedState: any) => {
  return configureStore({
    reducer: {
      rsu: rsuReducer,
      user: userReducer,
      wzdx: wzdxReducer,
      config: configReducer,
      intersection: intersectionReducer,
      adminAddOrganization: adminAddOrganizationReducer,
      adminAddRsu: adminAddRsuReducer,
      adminAddIntersection: adminAddIntersectionReducer,
      adminAddUser: adminAddUserReducer,
      adminEditOrganization: adminEditOrganizationReducer,
      adminEditRsu: adminEditRsuReducer,
      adminEditIntersection: adminEditIntersectionReducer,
      adminEditUser: adminEditUserReducer,
      adminOrganizationTab: adminOrganizationTabReducer,
      adminOrganizationTabUser: adminOrganizationTabUserReducer,
      adminOrganizationTabRsu: adminOrganizationTabRsuReducer,
      adminOrganizationTabIntersection: adminOrganizationTabIntersectionReducer,
      adminRsuTab: adminRsuTabReducer,
      adminIntersectionTab: adminIntersectionTabReducer,
      adminUserTab: adminUserTabReducer,
      adminNotificationTab: adminNotificationTabReducer,
      adminAddNotification: adminAddNotificationReducer,
      adminEditNotification: adminEditNotificationReducer,
      menu: menuReducer,
      intersectionMap: intersectionMapReducer,
      intersectionMapLayerStyle: intersectionMapLayerStyleReducer,
      dataSelector: dataSelectorReducer,
      map: mapSliceReducer,
      asn1Decoder: asn1DecoderSlice,
    },
    preloadedState,
    middleware: (getDefaultMiddleware) =>
      getDefaultMiddleware({
        thunk: true,
        serializableCheck: false,
        immutableCheck: false,
      }),
    devTools: true,
  })
}

type AppStore = ReturnType<typeof setupStore>
export type AppState = ReturnType<AppStore['getState']>
type AppThunk<ReturnType = void> = ThunkAction<ReturnType, AppState, unknown, Action>

export type RootState = ReturnType<ReturnType<typeof setupStore>['getState']>
