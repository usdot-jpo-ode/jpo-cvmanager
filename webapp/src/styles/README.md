# CVManager Theming

The CVManager natively supports 2 different themes: dark and light. Examples of each theme are shown below:

| Theme      | Map View                                                                       | Admin View                                                                         |
| ---------- | ------------------------------------------------------------------------------ | ---------------------------------------------------------------------------------- |
| Main Dark  | <img src="./screenshots/main_dark_map.png" alt="Main Dark Map" width="300"/>   | <img src="./screenshots/main_dark_admin.png" alt="Main Dark Admin" width="300"/>   |
| Main Light | <img src="./screenshots/main_light_map.png" alt="Main Light Map" width="300"/> | <img src="./screenshots/main_light_admin.png" alt="Main Light Admin" width="300"/> |

## Light vs Dark

The CVManager is configured to automatically change theme and icons based on the browser being in light vs dark mode. Each of the theme components, the base MUI theme and the icons, have different versions for light and dark. If you don't want the theme to change with the browser, simply set both to the same theme.

## Themes

As noted above, the two built-in themes are light and dark. Each of these themes has an existing MUI Theme object in [index.ts](./index.ts). These theme objects look as follows:

```ts
const themeMainDark = createTheme({
  //MUI createTheme call, initialize all un-specified values to theme defaults: https://mui.com/material-ui/customization/default-theme/
  components: {
    MuiTableHead: {
      // Custom table headers
      styleOverrides: {
        root: {
          backgroundColor: '#252525', // Custom table headers background color
          borderBottom: 'none',
          '& .MuiTableCell-root': {
            borderBottom: 'none',
            fontSize: '12px',
            fontWeight: 600,
            lineHeight: 1,
            letterSpacing: 0.5,
            textTransform: 'uppercase',
          },
          '& .MuiTableCell-paddingCheckbox': {
            paddingTop: 4,
            paddingBottom: 4,
          },
        },
      },
    },
  },
  palette: {
    mode: 'dark', // Should match general color scheme, slightly changes rendering
    primary: {
      main: '#b55e12', // Primary color, used in icon backgrounds, selected tabs, buttons, checkboxes, ...
      light: '#cecece', // Used as LightButton background (contact support and logout)
      dark: '#e37120', // Highlighted/hovered button backgrounds, used in smaller buttons, admin, toggles. Provides better text contrast than primary.main
      contrastText: '#fff', // Text color with contrast to primary.main and primary.dark colors
    },
    secondary: {
      main: '#0e2052', // Secondary color, used in snmpwalk backgrounds
      light: '#26329f', // Used in snmpwalk hovered backgrounds
      dark: '#0a0f3f', // Used as accent color for selected tab item, snmpwalk  backgrounds, and LightButton text color (contact support and logout)
      contrastText: '#fff',
    },
    error: {
      main: '#FD7C7C', // Text color in RSU firmware and RSU status tables, as well as ErrorMessageText. Background is background.paper
    },
    success: {
      main: '#90EE90', // Text color in RSU firmware and RSU status tables, as well as SuccessMessageText. Background is background.paper
    },
    text: {
      primary: '#ffffff', // Primary text color, used in most text-based components
      secondary: '#eb8841', // Used for Intersection page sub-titles
      disabled: '#acacac', // Used for disabled button text
    },
    custom: {
      mapLegendBackground: '#0e2052', // Used for horizontal tab backgrounds and map control/legend backgrounds
      tableHeaderBackground: '#252525', // Used for Admin table header background color, usually same as components.styleOverrides.root.backgroundColor
      tableErrorBackground: '#4d2e2e', // Background color of admin table rows when there is an error
      mapStyleFilePath: 'mapbox-styles/cdot-dark.json', // path to local mapbox map style.json
      mapStyleHasTraffic: false, // Whether the style contains mapbox traffic and incident layers
    },
    divider: '#111', // Used in dividers (<Divider /> components)
    background: {
      paper: '#333', // Used in table body backgrounds, in-page header backgrounds, and vertical tab background. Also used in Help page background
      default: '#1c1d1f', // Used as default in page backgrounds
    },
  },
})
```

## Icons

The main icon/logo in the CVManager is used in 2 places: the after-login loading/error page, and the main dashboard at the top-left.

The CVManager allows customization of icons when building the application through docker. This is done by hosting the icons as volumes in the running webapp image, to the path container /usr/share/nginx/html/icons/logo_light.png and /usr/share/nginx/html/icons/logo_dark.png (path in repo is webapp/public/icons/logo_light and webapp/public/icons/logo_dark). This is already done in [docker-compose.yml](../../../docker-compose.yml#L79), and is set through the env vars WEBAPP_LOGO_PNG_ROOT_FILE_PATH_LIGHT and WEBAPP_LOGO_PNG_ROOT_FILE_PATH_DARK. These should each be set to the path of a .png image, which can be inside or outside of this repository.

Icons will be locked to 90px high when displayed.
