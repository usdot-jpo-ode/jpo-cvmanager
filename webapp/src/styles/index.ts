import { alpha, createTheme, Theme } from '@mui/material'
import './fonts/museo-slab.css'
import '../App.css'

declare module '@mui/material/styles' {
  interface PaletteOptions {
    custom: {
      mapLegendBackground: string
      tableHeaderBackground: string
      tableErrorBackground: string
      mapStyleFilePath: string
      mapStyleHasTraffic: boolean
      mapMenuItemBackgroundSelected: string
      mapMenuItemBorderSelected: string
      mapMenuItemHoverSelected: string
      mapMenuItemHoverUnselected: string
      mapMenuBackground: string
      intersectionMapAccordionExpanded: string
      intersectionMapButtonHover: string
      rowActionIcon: string
    }
  }
  interface Palette {
    custom: {
      mapLegendBackground: string
      tableHeaderBackground: string
      tableErrorBackground: string
      mapStyleFilePath: string
      mapStyleHasTraffic: boolean
      mapMenuItemBackgroundSelected: string
      mapMenuItemBorderSelected: string
      mapMenuItemHoverSelected: string
      mapMenuItemHoverUnselected: string
      mapMenuBackground: string
      intersectionMapAccordionExpanded: string
      intersectionMapButtonHover: string
      rowActionIcon: string
    }
  }
}

export const headerTabHeight = 119

export const testTheme = createTheme({
  palette: {
    custom: {
      mapLegendBackground: '#000000',
      tableHeaderBackground: '#000000',
      tableErrorBackground: '#000000',
      mapStyleFilePath: 'mapbox-styles/cdot-dark.json',
      mapStyleHasTraffic: true,
      mapMenuItemBackgroundSelected: '#000000',
      mapMenuItemBorderSelected: '#000000',
      mapMenuItemHoverSelected: '#000000',
      mapMenuItemHoverUnselected: '#000000',
      mapMenuBackground: '#000000',
      intersectionMapAccordionExpanded: '#000000',
      intersectionMapButtonHover: '#000000',
      rowActionIcon: '#000000',
    },
  },
})

// Please note that the light theme is currently not being maintained.
// Light Theme - https://www.realtimecolors.com/?colors=0a1424-e7eef8-213e73-7978d9-4431af&fonts=Inter-Inter
// --text: #0a1424;
// --background: #e7eef8;
// --primary: #213e73;
// --secondary: #7978d9;
// --accent: #4431af;
const themeMainLight = createTheme({
  cssVariables: true,
  typography: {
    fontFamily: '"Trebuchet MS", Arial, Helvetica, sans-serif',
    h1: {
      fontFamily: '"museo-slab" Arial Helvetica Sans-Serif',
    },
    h2: {
      fontFamily: '"museo-slab" Arial Helvetica Sans-Serif',
    },
    h3: {
      fontFamily: '"museo-slab" Arial Helvetica Sans-Serif',
    },
    h4: {
      fontFamily: '"museo-slab" Arial Helvetica Sans-Serif',
    },
    h5: {
      fontFamily: '"museo-slab" Arial Helvetica Sans-Serif',
    },
    h6: {
      fontFamily: '"museo-slab" Arial Helvetica Sans-Serif',
    },
  },
  components: {
    MuiTableHead: {
      styleOverrides: {
        root: {
          backgroundColor: '#F3F4F6',
          '.MuiTableCell-root': {
            color: '#374151',
          },
          borderBottom: 'none',
          '& .MuiTableCell-root': {
            borderBottom: 'none',
            fontSize: '12px',
            fontWeight: 600,
            lineHeight: 1,
            letterSpacing: 0.5,
            textTransform: 'uppercase',
          },
          '& .MuiTableCell-paddingCheckbox': {
            paddingTop: 4,
            paddingBottom: 4,
          },
        },
      },
    },
    MuiPaper: {
      styleOverrides: {
        root: {
          backgroundImage: 'unset',
        },
      },
    },
    MuiDivider: {
      styleOverrides: {
        root: {
          borderColor: '#061731',
        },
      },
    },
  },
  palette: {
    mode: 'light',
    primary: {
      main: '#6f69e0',
      light: '#2e3574',
      dark: '#a7a3e7',
      contrastText: '#0b041b',
    },
    secondary: {
      main: '#53aaf1',
      light: '#114875',
      dark: '#92bcde',
      contrastText: '#121212',
    },
    error: {
      main: '#713737',
    },
    success: {
      main: '#315131',
    },
    text: {
      primary: '#121828',
      secondary: '#061731',
      disabled: 'rgba(55, 65, 81, 0.48)',
    },
    divider: '#8f8f8f',
    background: {
      default: '#ececec',
      paper: '#FFFFFF',
    },
    custom: {
      mapLegendBackground: '#c8cfda',
      tableHeaderBackground: '#F3F4F6',
      tableErrorBackground: '#fdc7c7',
      mapStyleFilePath: 'mapbox-styles/main-light.json',
      mapStyleHasTraffic: true,
      mapMenuBackground: '#e0e0e0',
      mapMenuItemBackgroundSelected: '#c4c2c2',
      mapMenuItemBorderSelected: '1px solid black',
      mapMenuItemHoverSelected: '#a19f9f',
      mapMenuItemHoverUnselected: '#ffffff',
      intersectionMapAccordionExpanded: '#d7d7d7',
      intersectionMapButtonHover: '#d7d7d7',
      rowActionIcon: '#9DBDD3',
    },
  },
})

// --text: #dbe5f5;
// --background: #1b1d1f;
// --primary: #4383ad;
// --secondary: #dbe5f5;
// --accent: #614fcd;
const themeMainDark = createTheme({
  cssVariables: true,
  typography: {
    fontFamily: '"Trebuchet MS", Arial, Helvetica, sans-serif',
    h1: {
      fontFamily: '"museo-slab" Arial Helvetica Sans-Serif',
    },
    h2: {
      fontFamily: '"museo-slab" Arial Helvetica Sans-Serif',
    },
    h3: {
      fontFamily: '"museo-slab" Arial Helvetica Sans-Serif',
    },
    h4: {
      fontFamily: '"museo-slab" Arial Helvetica Sans-Serif',
    },
    h5: {
      fontFamily: '"museo-slab" Arial Helvetica Sans-Serif',
    },
    h6: {
      fontFamily: '"museo-slab" Arial Helvetica Sans-Serif',
    },
  },
  components: {
    MuiTableHead: {
      styleOverrides: {
        root: {
          backgroundColor: '#252525',
          borderBottom: 'none',
          '& .MuiTableCell-root': {
            borderBottom: 'none',
            fontSize: '12px',
            fontWeight: 600,
            lineHeight: 1,
            letterSpacing: 0.5,
            textTransform: 'uppercase',
          },
          '& .MuiTableCell-paddingCheckbox': {
            paddingTop: 4,
            paddingBottom: 4,
          },
        },
      },
    },
    MuiPaper: {
      styleOverrides: {
        root: {
          backgroundImage: 'unset',
        },
      },
    },
    MuiDivider: {
      styleOverrides: {
        root: {
          borderColor: '#c7c7c7',
        },
      },
    },
  },
  palette: {
    mode: 'dark',
    primary: {
      main: '#3B7BA5',
      light: '#51a2d6',
      dark: '#3B7BA54D',
      contrastText: '#fff',
    },
    secondary: {
      main: '#dbe5f5',
      light: '#535297',
      dark: '#161563',
      contrastText: '#FAFAFA',
    },
    error: {
      light: '#FF6D57',
      main: '#E94F3766',
      dark: '#6e312a',
    },
    success: {
      light: '#75BD27',
      main: '#A0D36466',
      dark: '#9e0e0e',
    },
    text: {
      primary: '#FFFFFF',
      secondary: '#c7c7c7',
      disabled: '#acacac',
    },
    info: {
      main: '#dbe5f5',
    },
    custom: {
      mapLegendBackground: '#1b1d1f',
      tableHeaderBackground: '#252525',
      tableErrorBackground: '#4d2e2e',
      mapStyleFilePath: 'mapbox-styles/main-dark.json',
      mapStyleHasTraffic: true,
      mapMenuBackground: '#3c3c3c',
      mapMenuItemBackgroundSelected: '#333333',
      mapMenuItemBorderSelected: '1px solid black',
      mapMenuItemHoverSelected: '#333333',
      mapMenuItemHoverUnselected: '#575757',
      intersectionMapAccordionExpanded: '#2E2F31',
      intersectionMapButtonHover: '#2D5F7F',
      rowActionIcon: '#9DBDD3',
    },
    background: {
      paper: '#1b1d1f',
      default: '#333333',
    },
  },
})

// All available themes
export const THEMES = {
  light: themeMainLight,
  dark: themeMainDark,
}

export const getCurrentTheme = (isDarkTheme: boolean, lightThemeName: string, darkThemeName: string): Theme => {
  // Warnings to user if theme names are not known
  if (darkThemeName && !THEMES[darkThemeName]) {
    console.warn(`Unknown dark theme name: ${darkThemeName}`)
  }
  if (lightThemeName && !THEMES[lightThemeName]) {
    console.warn(`Unknown light theme name: ${lightThemeName}`)
  }
  if (isDarkTheme) {
    return THEMES[darkThemeName] ?? THEMES.dark
  } else {
    return THEMES[lightThemeName] ?? THEMES.light
  }
}
