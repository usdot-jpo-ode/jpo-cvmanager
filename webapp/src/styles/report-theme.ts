import { createTheme } from '@mui/material'

export const ReportTheme = createTheme({
  typography: {
    h1: {
      color: 'black',
    },
    h2: {
      color: 'black',
    },
    h3: {
      color: 'black',
    },
    h4: {
      color: 'black',
    },
    h5: {
      color: 'black',
    },
    h6: {
      color: 'black',
    },
    body1: {
      color: 'black',
    },
    body2: {
      color: 'black',
    },
  },
  palette: {
    mode: 'light',
    primary: {
      main: '#000000',
      light: '#000000',
      dark: '#000000',
      contrastText: '#FFFFFF',
    },
    secondary: {
      main: '#000000',
      light: '#000000',
      dark: '#000000',
      contrastText: '#FFFFFF',
    },
    error: {
      main: '#ad2626',
    },
    success: {
      main: '#2e942e',
    },
    text: {
      primary: '#030303',
      secondary: '#030303',
      disabled: 'rgba(66, 66, 66, 0.48)',
    },
    divider: '#575757',
    background: {
      default: '#ffffff',
      paper: '#FFFFFF',
    },
    custom: {
      mapLegendBackground: '#ffffff',
      tableHeaderBackground: '#ffffff',
      tableErrorBackground: '#ffffff',
      mapStyleFilePath: 'mapbox-styles/main-light.json',
      mapStyleHasTraffic: true,
      mapMenuBackground: '#ffffff',
      mapMenuItemBackgroundSelected: '#ffffff',
      mapMenuItemBorderSelected: '1px solid black',
      mapMenuItemHoverSelected: '#a19f9f',
      mapMenuItemHoverUnselected: '#ffffff',
      rowActionIcon: '#9DBDD3',
    },
  },
})
