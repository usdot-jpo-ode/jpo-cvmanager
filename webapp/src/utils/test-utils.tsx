/**
 * This function searches for the every somewhat random/chaotic class name and property that cause snapshot tests to be inconsistent.
 * This current list includes MUI classes (css-*) and aria-invalid attributes.
 *
 * @ex :
 * ```
 * const { container } = render(<Component />);
 *
 * replaceChaoticIds(container);
 * ```
 *
 * @param container The HTMLElement node to search for SSR ids
 */
function replaceChaoticIds(container: HTMLElement) {
  const props = [
    {
      selector: 'class',
      updateFunc: (val: string) => val.replace(/css-[0-9a-z]*?-/g, 'css-mocked-'),
    },
  ]

  container.querySelectorAll('*').forEach((item) => {
    props.forEach((prop) => {
      if (item.getAttribute(prop.selector)) {
        item.setAttribute(prop.selector, prop.updateFunc(item.getAttribute(prop.selector)))
      }
    })
  })
  container.querySelectorAll('input[aria-invalid]').forEach((item) => {
    item.removeAttribute('aria-invalid')
  })
  return container
}

import React from 'react'
import { LocalizationProvider } from '@mui/x-date-pickers/LocalizationProvider'
import { AdapterDayjs } from '@mui/x-date-pickers/AdapterDayjs'
import dayjs from 'dayjs'
import utc from 'dayjs/plugin/utc'
import timezone from 'dayjs/plugin/timezone'

// Extend dayjs with timezone and UTC plugins
dayjs.extend(utc)
dayjs.extend(timezone)

// Mock component to set timezone
const MockLocalizationProvider = ({ children }) => {
  dayjs.tz.setDefault('America/Denver')
  return <LocalizationProvider dateAdapter={AdapterDayjs}>{children}</LocalizationProvider>
}

export { replaceChaoticIds, MockLocalizationProvider }
