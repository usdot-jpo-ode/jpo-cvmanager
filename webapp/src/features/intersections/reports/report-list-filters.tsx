import React, { useState, useEffect } from 'react'
import PropTypes from 'prop-types'
import { Box, Button, CircularProgress, Drawer, IconButton, Stack, TextField, Typography } from '@mui/material'
import { styled } from '@mui/material/styles'
import { DateTimePicker } from '@mui/x-date-pickers/DateTimePicker'
import { Close } from '@mui/icons-material'
import { AdapterDateFns } from '@mui/x-date-pickers/AdapterDateFns'

import { LocalizationProvider } from '@mui/x-date-pickers/LocalizationProvider'
import dayjs from 'dayjs'

const FiltersDrawerDesktop = styled(Drawer)({
  flexShrink: 0,
  width: 380,
  '& .MuiDrawer-paper': {
    position: 'relative',
    width: 380,
  },
})

const FiltersDrawerMobile = styled(Drawer)({
  maxWidth: '100%',
  width: 380,
  '& .MuiDrawer-paper': {
    height: 'calc(100% - 64px)',
    maxWidth: '100%',
    top: 64,
    width: 380,
  },
})

export type ReportListFilter = {
  startDate: Date | null
  endDate: Date | null
}

interface ReportListFiltersProps {
  filters: ReportListFilter
  onChange: (filters: ReportListFilter) => void
  onClose: () => void
  open: boolean
  loading: boolean
  containerRef: any
  setOpenReportGenerationDialog: (open: boolean) => void
}

export const ReportListFilters = (props: ReportListFiltersProps) => {
  const { containerRef, filters, onChange, onClose, open, loading, setOpenReportGenerationDialog, ...other } = props
  const [currentFilters, setCurrentFilters] = useState(filters)
  const [filtersValid, setFiltersValid] = useState([true, ''])

  useEffect(() => {
    updateFiltersValid()
  }, [currentFilters])

  const startDateChange = (date: Date | null) => {
    setCurrentFilters({ ...currentFilters, startDate: date })
  }

  const endDateChange = (date: Date | null) => {
    setCurrentFilters({ ...currentFilters, endDate: date })
  }

  const updateFilters = () => {
    onChange?.(currentFilters)
  }

  const updateFiltersValid = () => {
    let filtersValidLocal = true
    const reasons: string[] = []
    if (currentFilters.startDate === null || currentFilters.endDate === null) {
      filtersValidLocal = false
      reasons.push('Start Date and End Date must be set')
    } else if (currentFilters.startDate >= currentFilters.endDate) {
      filtersValidLocal = false
      reasons.push('Start Date must be before End Date')
    }
    // TODO: validate dates
    setFiltersValid([filtersValidLocal, reasons.join('\n')])
    if (filtersValidLocal) {
      onChange?.(currentFilters)
    }
  }

  const content = (
    <Box
      sx={{
        pb: 1,
        pt: 1,
        px: 3,
      }}
    >
      <Box
        sx={{
          mb: 2,
        }}
      >
        <IconButton onClick={onClose}>
          <Close fontSize="small" />
        </IconButton>
      </Box>
      <Typography color="textSecondary" sx={{ mt: 3 }} variant="subtitle2">
        Issue date
      </Typography>
      <Stack spacing={2} sx={{ mt: 2 }}>
        <LocalizationProvider dateAdapter={AdapterDayjs}>
          <DateTimePicker
            label="From"
            onChange={(e) => startDateChange(e?.toDate())}
            renderInput={(inputProps) => <TextField {...inputProps} />}
            disabled={loading}
          />
          <DateTimePicker
            label="To"
            onChange={(e) => endDateChange(e?.toDate())}
            renderInput={(inputProps) => <TextField {...inputProps} />}
            disabled={loading}
          />
        </LocalizationProvider>
      </Stack>
      {!filtersValid[0] && (
        <Typography color="red" sx={{ mt: 3 }} variant="subtitle2">
          Invalid filters: {filtersValid[1]}
        </Typography>
      )}
      <Box sx={{ m: 1, position: 'relative', display: 'flex', alignItems: 'center' }}>
        {loading && (
          <CircularProgress
            size={24}
            sx={{
              position: 'absolute',
              top: '50%',
              left: '50%',
              marginTop: '-12px',
              marginLeft: '-12px',
            }}
          />
        )}
      </Box>
      <Button
        component="a"
        onClick={() => {
          setOpenReportGenerationDialog(true)
        }}
      >
        Generate Manual Report
      </Button>
    </Box>
  )

  return (
    <FiltersDrawerDesktop
      anchor="left"
      open={open}
      SlideProps={{ container: containerRef?.current }}
      variant="persistent"
      {...other}
    >
      {content}
    </FiltersDrawerDesktop>
  )
}

ReportListFilters.propTypes = {
  containerRef: PropTypes.any,
  filters: PropTypes.object,
  onChange: PropTypes.func,
  onClose: PropTypes.func,
  open: PropTypes.bool,
  loading: PropTypes.bool,
}
