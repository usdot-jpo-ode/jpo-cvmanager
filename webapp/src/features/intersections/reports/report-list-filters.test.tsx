import React from 'react'
import { render } from '@testing-library/react'
import { ReportListFilters } from './report-list-filters'
import { Provider } from 'react-redux'
import { ThemeProvider } from '@mui/material'
import { testTheme } from '../../../styles'
import { setupStore } from '../../../store'
import { replaceChaoticIds } from '../../../utils/test-utils'

it('should take a snapshot', () => {
  const { container } = render(
    <ThemeProvider theme={testTheme}>
      <Provider store={setupStore({})}>
        <ReportListFilters
          open={true}
          onClose={() => {}}
          filters={{ startDate: new Date('2025-04-07'), endDate: new Date('2025-04-08') }}
          onChange={() => {}}
          loading={false}
          containerRef={() => {}}
          setOpenReportGenerationDialog={() => {}}
        />
      </Provider>
    </ThemeProvider>
  )

  expect(replaceChaoticIds(container)).toMatchSnapshot()
})
