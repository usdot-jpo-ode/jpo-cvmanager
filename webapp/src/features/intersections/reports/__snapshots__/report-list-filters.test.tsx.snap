// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should take a snapshot 1`] = `
<div>
  <div
    class="MuiBox-root css-t0051x"
  >
    <div
      class="MuiBox-root css-1ebprri"
    >
      <div
        class="MuiFormControl-root MuiTextField-root css-mocked-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-mocked-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="true"
          for="mui-1"
          id="mui-1-label"
        >
          From
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedEnd css-mocked-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            autocomplete="off"
            class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd css-mocked-MuiInputBase-input-MuiOutlinedInput-input"
            id="mui-1"
            inputmode="text"
            placeholder="MM/DD/YYYY hh:mm aa"
            type="text"
            value="04/06/2025 06:00 PM"
          />
          <div
            class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-mocked-MuiInputAdornment-root"
          >
            <button
              aria-label="Choose date, selected date is Apr 6, 2025"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-mocked-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-mocked-MuiSvgIcon-root"
                data-testid="CalendarIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M17 12h-5v5h5v-5zM16 1v2H8V1H6v2H5c-1.11 0-1.99.9-1.99 2L3 19c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2h-1V1h-2zm3 18H5V8h14v11z"
                />
              </svg>
            </button>
          </div>
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-mocked-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-w1u3ce"
            >
              <span>
                From
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        class="MuiFormControl-root MuiTextField-root css-mocked-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-mocked-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="true"
          for="mui-2"
          id="mui-2-label"
        >
          To
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedEnd css-mocked-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            autocomplete="off"
            class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd css-mocked-MuiInputBase-input-MuiOutlinedInput-input"
            id="mui-2"
            inputmode="text"
            placeholder="MM/DD/YYYY hh:mm aa"
            type="text"
            value="04/07/2025 06:00 PM"
          />
          <div
            class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-mocked-MuiInputAdornment-root"
          >
            <button
              aria-label="Choose date, selected date is Apr 7, 2025"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-mocked-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-mocked-MuiSvgIcon-root"
                data-testid="CalendarIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M17 12h-5v5h5v-5zM16 1v2H8V1H6v2H5c-1.11 0-1.99.9-1.99 2L3 19c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2h-1V1h-2zm3 18H5V8h14v11z"
                />
              </svg>
            </button>
          </div>
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-mocked-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-w1u3ce"
            >
              <span>
                To
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
    <div
      class="MuiBox-root css-gjqw5t"
    />
    <a
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary museo-slab capital-case css-mocked-MuiButtonBase-root-MuiButton-root"
      role="button"
      tabindex="0"
    >
      Generate Manual Report
    </a>
  </div>
</div>
`;
